src/main/java/org/springframework/samples/petclinic/rest/importcsv/ImportCSV.java

```diff
@@ -33,107 +40,153 @@
+
+        return entries;
+    }
+
+
+
+    private Pet convert_string_list_to_pet( List<String> list )
+    {
+
+        Pet pet = new Pet();
+        pet.setName(list.get(0));
+        pet.setBirthDate(convert_birth_date_from_string(list.get(1)));
+        pet.setType(search_for_pet_type_in_list(list.get(2)));
 
-            } else {
-                clinicService.savePet(pet);
+        // ADD EXCEPTION FOR NON-UNIQUE OWNERS !
+        List<Owner> matchingOwners = getMatchingOwners(list.get(3));
+        if ( isUniqueOwner(matchingOwners) )
+        {
+            //OLD CODE: pet.setOwner(matchingOwners.iterator().next());
+            pet.setOwner(matchingOwners.get(0));
+        }
+
+        return pet;
+    }
+
+    private Date convert_birth_date_from_string(String birth_date_string)
+    {
+        Date date = new Date();
+
+        SimpleDateFormat simple_date_format = new SimpleDateFormat("yyyy-MM-dd");
+
+        try {
+            date = simple_date_format.parse(birth_date_string);
+        }
+        catch (ParseException e) {
+            HttpHeaders headers = new HttpHeaders();
+            headers.add("errors", "date " + birth_date_string + " not valid");
+
+        }
+
+        return date;
+
+    }
+
+    private ArrayList<PetType> get_pet_types_from_server()
+    {
+        ArrayList<PetType> pet_type_list = (ArrayList<PetType>) clinicService.findPetTypes();
+        return pet_type_list;
+    }
+
+    private PetType search_for_pet_type_in_list(String entry)
+    {
+        PetType pet_type = new PetType();
+
+        ArrayList<PetType> pet_type_list = get_pet_types_from_server();
+
+        for (int petIndex = 0; petIndex < pet_type_list.size(); petIndex++)
+        {
+            if (pet_type_list.get(petIndex).getName().toLowerCase().equals(entry))
+            {
+                pet_type= pet_type_list.get(petIndex);
+                break;
             }
-            i++;
+        }
 
-            pets.add(pet);
+        return pet_type;
+    }
 
-        } while (i < csv.length() && pet != null);
+    private boolean isUniqueOwner(List<Owner> matchingOwners)
+    {
+        if (matchingOwners.size() == 1) {
+          return true;
+        } else if (matchingOwners.size() == 0) {
+            HttpHeaders headers = new HttpHeaders();
+            headers.add("errors", "Owner not found");
+            return false;
+        } else { //if (matchingOwners.size() > 1)
+            HttpHeaders headers = new HttpHeaders();
+            headers.add("errors", "Owner not unique");
+            return false;
+        }
+    }
 
-        return new ResponseEntity<List<Pet>>(pets, HttpStatus.OK);
+    private List<Owner> getMatchingOwners(String entry)
+    {
+      List<Owner> matchingOwners = clinicService.findAllOwners()
+          .stream()
+          .filter(o -> o.getLastName().equals(entry))
+          .collect(Collectors.toList());
+      return matchingOwners;
     }
+
+    private boolean isSamePet(Pet pet1, Pet pet2)
```

Comment:
I like that you did not use more than 2 indentations (except in this method `isSamePet` but you would have improved this probably with more time)